crates/rpc/rpc/src/eth/api/fees.rs

```diff
@@ -38,167 +37,148 @@ where
+
+            // Percentiles were specified, so we need to collect reward percentile ino
+            if let Some(percentiles) = &reward_percentiles {
+                rewards.push(self.calculate_reward_percentiles(&percentiles, header)?);
             }
         }
 
-        // get the first block in the range from the db
-        let oldest_block_hash =
-            self.inner.provider.block_hash(start_block)?.ok_or(EthApiError::UnknownBlockNumber)?;
-
-        // Set the hash in cache items if the block is present in the cache
-        if let Some(cache_item) = fee_history_cache_items.get_mut(&start_block) {
-            cache_item.hash = Some(oldest_block_hash);
-        }
-
-        if let Some(cache_item) = fee_history_cache.get_mut(&start_block) {
-            cache_item.hash = Some(oldest_block_hash);
-        }
-
-        // `fee_history_cache_items` now contains full requested block range (populated from both
-        // cache and database), so we can iterate over it in order and populate the response fields
-        let base_fee_per_gas =
-            fee_history_cache_items.values().map(|item| item.base_fee_per_gas).collect();
-
-        let mut gas_used_ratio: Vec<f64> =
-            fee_history_cache_items.values().map(|item| item.gas_used_ratio).collect();
-
-        let mut rewards: Vec<Vec<_>> =
-            fee_history_cache_items.values().filter_map(|item| item.reward.clone()).collect();
-
-        // gasUsedRatio doesn't have data for next block in this case the last block
-        gas_used_ratio.pop();
-        rewards.pop();
+        // The spec states that `base_fee_per_gas` "[..] includes the next block after the newest of
+        // the returned range, because this value can be derived from the newest block"
+        //
+        // The unwrap is safe since we checked earlier that we got at least 1 header.
+        let last_header = headers.last().unwrap();
+        base_fee_per_gas.push(U256::from(calculate_next_block_base_fee(
+            last_header.gas_used,
+            last_header.gas_limit,
+            last_header.base_fee_per_gas.unwrap_or_default(),
+        )));
 
         Ok(FeeHistory {
             base_fee_per_gas,
             gas_used_ratio,
             oldest_block: U256::from(start_block),
-            reward: Some(rewards),
+            reward: reward_percentiles.map(|_| rewards),
         })
     }
+
+    // todo: docs
+    fn calculate_reward_percentiles(
+        &self,
+        percentiles: &[f64],
+        header: &Header,
+    ) -> Result<Vec<U256>, EthApiError> {
+        let Some(receipts) =
+            self.inner.provider.receipts_by_block(header.number.into())? else {
+            // If there are no receipts, then we do not have all info on the block
+            return Err(EthApiError::InvalidBlockRange)
+        };
+        let Some(mut transactions): Option<Vec<_>> = self
+            .inner
+            .provider
+            .transactions_by_block(header.number.into())?
```

Comment:
these should make use of the async caching layer I think

Reply:
how do I do that?